packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_Utilities_MP_Vector.hpp

```diff
@@ -241,269 +241,65 @@ namespace Stokhos {
-    Teuchos::RCP< Tpetra::MultiVector<typename Storage::value_type,LocalOrdinal,GlobalOrdinal,Node> > flat_mv = create_flat_vector_view(mv, flat_map);
-    return flat_mv->getVector(0);
-  }
-
-  // Create a flattened vector by unrolling the MP::Vector scalar type.  The
-  // returned vector is a view of the original.  This version creates the
-  // map if necessary
-  template <typename Storage, typename LocalOrdinal, typename GlobalOrdinal,
-            typename Node>
-  Teuchos::RCP< const Tpetra::Vector<typename Storage::value_type,
-                                     LocalOrdinal,GlobalOrdinal,Node> >
-  create_flat_vector_view(
-    const Tpetra::Vector<Sacado::MP::Vector<Storage>,
-                         LocalOrdinal,GlobalOrdinal,Node>& vec,
-    Teuchos::RCP< const Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> >& flat_map) {
+    const Tpetra::MultiVector<Sacado::MP::Vector<Storage>, LocalOrdinal, GlobalOrdinal, Node>& vec,
+          Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>>& flat_map
+  ) {
     if (flat_map == Teuchos::null) {
       const LocalOrdinal mp_size = Storage::static_size;
       flat_map = create_flat_map(*(vec.getMap()), mp_size);
     }
-    const Teuchos::RCP< const Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> > const_flat_map = flat_map;
+    const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> const_flat_map = flat_map;
     return create_flat_vector_view(vec, const_flat_map);
   }
 
-  // Create a flattened vector by unrolling the MP::Vector scalar type.  The
-  // returned vector is a view of the original
-  template <typename Storage, typename LocalOrdinal, typename GlobalOrdinal,
-            typename Node>
-  Teuchos::RCP< Tpetra::Vector<typename Storage::value_type,
-                               LocalOrdinal,GlobalOrdinal,Node> >
+  template <typename Storage, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+  Teuchos::RCP<Tpetra::Vector<typename Storage::value_type, LocalOrdinal, GlobalOrdinal, Node>>
   create_flat_vector_view(
-    Tpetra::Vector<Sacado::MP::Vector<Storage>,
-                   LocalOrdinal,GlobalOrdinal,Node>& vec,
-    const Teuchos::RCP< const Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> >& flat_map) {
-    Tpetra::MultiVector<Sacado::MP::Vector<Storage>,LocalOrdinal,GlobalOrdinal,Node>& mv = vec;
-    Teuchos::RCP< Tpetra::MultiVector<typename Storage::value_type,LocalOrdinal,GlobalOrdinal,Node> > flat_mv = create_flat_vector_view(mv, flat_map);
-    return flat_mv->getVectorNonConst(0);
+    const Tpetra::Vector<Sacado::MP::Vector<Storage>, LocalOrdinal, GlobalOrdinal, Node>& vec,
+    const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>>& flat_map
+  ) {
+    return create_flat_vector_view(
+      dynamic_cast<const Tpetra::MultiVector<Sacado::MP::Vector<Storage>, LocalOrdinal, GlobalOrdinal, Node>&>(vec),
```

Comment:
This could be a ``static_cast`` since Vector is a subclass of MultiVector

Reply:
But this is a very minor thing, if you're not making any other changes for the autotester then don't worry about it

Reply:
Thanks, @brian-kelley ! I've just changed it into a `static_cast`.